clang/lib/Sema/SemaBounds.cpp

```diff
@@ -4713,168 +4760,248 @@ namespace {
+      BoundsExpr *AdjustedSrcBounds = SrcBounds;
+      if (!M) {
+        // If LValue has target bounds, then the observed bounds of LValue
+        // are already SrcBounds adjusted to account for any use of LValue.
+        if (HasTargetBounds)
+          AdjustedSrcBounds = State.ObservedBounds[LValueAbstractSet];
+        else
+          AdjustedSrcBounds = ReplaceLValueInBounds(SrcBounds, LValue,
+                                                    OriginalValue, CSS);
       }
+      
 
-      // Adjust SrcBounds to account for any uses of V and, if V has declared
-      // bounds, record the updated observed bounds for V.
-      BoundsExpr *AdjustedSrcBounds = ReplaceVariableInBounds(SrcBounds, V, OriginalValue, CSS);
-      if (DeclaredBounds)
-        State.ObservedBounds[VariableAbstractSet] = AdjustedSrcBounds;
-
-      // Record that E updates the observed bounds of V.
-      if (DeclaredBounds)
-        State.BlameAssignments[VariableAbstractSet] = E;
+      // Record that E updates the observed bounds of LValue.
+      if (HasTargetBounds)
+        State.BlameAssignments[LValueAbstractSet] = E;
 
       // If the initial source bounds were not unknown, but they are unknown
-      // after replacing uses of V, then the assignment to V caused the
-      // source bounds (which are the observed bounds for V) to be unknown.
-      if (DeclaredBounds) {
+      // after replacing uses of LValue, then the assignment to LValue caused
+      // the source bounds (which are the observed bounds for LValue) to be
+      // unknown.
+      if (HasTargetBounds) {
         if (!SrcBounds->isUnknown() && AdjustedSrcBounds->isUnknown())
-          State.LostVariables[VariableAbstractSet] = std::make_pair(SrcBounds, V);
+          State.LostLValues[LValueAbstractSet] =
+            std::make_pair(SrcBounds, LValue);
       }
 
-      // Adjust EquivExprs to account for any uses of V in PrevState.EquivExprs.
+      return AdjustedSrcBounds;
+    }
+
+    // UpdateEquivExprsAfterAssignment updates the set EquivExprs of set of
+    // equivalent expressions after the expression LValue is modified via
+    // an assignment.
+    //
+    // OriginalValue is the value of LValue before the assignment.
+    // If OriginalValue is non-null, each expression in EquivExprs has all
+    // occurrences of LValue replaced with OriginalValue.
+    // If OriginalValue is null, each expression in EquivExprs that uses the
+    // value of LValue is removed from EquivExprs.
+    void UpdateEquivExprsAfterAssignment(Expr *LValue, Expr *OriginalValue,
+                                         CheckedScopeSpecifier CSS,
+                                         CheckingState &State) {
+      const EquivExprSets PrevEquivExprs = State.EquivExprs;
       State.EquivExprs.clear();
-      for (auto I = PrevState.EquivExprs.begin(); I != PrevState.EquivExprs.end(); ++I) {
+      for (auto I = PrevEquivExprs.begin(); I != PrevEquivExprs.end(); ++I) {
```

Comment:
Since `PrevEquivExprs` is not modified inside the loop we don't need to re-evaluate `PrevEquivExprs.end()` on each loop iteration. For better readability, this loop can be written in a range-based format instead of using iterators. Same for the `InnerList` loop below.

Note: This is only a suggestion and does not need to be done in the current PR.

Reply:
There are several instances of this in SemaBounds.cpp. Created issue #1077 to track this cleanup work item.